### PR TITLE
fix: content budget math

### PR DESCRIPTION
**Pull request by rachmlenig**
### **PR Type**
Bug fix, Tests

___

### **Description**
- Fix context budget math for accurate token accounting and completion cap

- Add native tool rendering support with proper context validation

- Implement request-aware budget allocation for completion tokens

- Add regression tests and session cache invalidation on config changes

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["Chat Request"] --> B["Prepare Messages"]
  B --> C["Check Tool Rendering"]
  C -->|Native Jinja2| D["Count Rendered Prompt"]
  C -->|Prompt Injection| E["Inject Tools into Messages"]
  D --> F["Validate Context Budget"]
  E --> F
  F -->|Over Budget| G["Truncate or Error"]
  F -->|Within Budget| H["Cap Completion Tokens"]
  G --> I["Return Response"]
  H --> I
```

<details><summary><h3>File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td><strong>gguf_language_model.py</strong><dd><code>Add context validation method for tool rendering</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/779/files#diff-aa3d47952f93f3f6d338873bbc86b4f9eb2700b79e17fd3e73e5a2016fc37064">+26/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Bug fix</strong></td><td><details><summary>3 files</summary><table>
<tr>
  <td><strong>service.py</strong><dd><code>Fix budget math and native tool context counting</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/779/files#diff-0d5bb320d0590e4a8fb439fd35b07be371bae7d283b4862979f25fbaad05be14">+167/-37</a></td>

</tr>

<tr>
  <td><strong>context_manager.py</strong><dd><code>Fix budget calculation for small context windows</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/779/files#diff-a4f00b61608b77a18c221a50fd4879d10defe57435a03ce447a8ab996f97b06d">+26/-12</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>projects.py</strong><dd><code>Invalidate session cache on config modification</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/779/files#diff-ab4cacdb0344ceca710f889b0d839b684ea3bd780df3f6125661cf30950b287a">+12/-1</a>&nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>5 files</summary><table>
<tr>
  <td><strong>test_context_manager.py</strong><dd><code>Add regression tests for budget edge cases</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/779/files#diff-ededdc49652edf7652945c744358e031689669cd77b3ccd51105863adce451f4">+22/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>test_gguf_model.py</strong><dd><code>Add tests for context validation method</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/779/files#diff-7527a7beefa22a8f08b97a9c2fba080f4c6b9f6841c812ecf265fac15a7e067e">+54/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>test_tool_calling_e2e.py</strong><dd><code>Add comprehensive tool and budget validation tests</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/779/files#diff-60e86dece7958bc1f51577eee8850c405dae08e7bfc6ace126ab4d8625437853">+233/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>test_chat_api_baseline.py</strong><dd><code>Add test for session invalidation on config change</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/779/files#diff-f3b88a0b8178c54a8a98d1fe619aa3ab32c390783b4dc15870e3554dedb0eaf1">+39/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>test_projects_api.py</strong><dd><code>Update session record initialization</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/779/files#diff-77293a756c1d216c90486e2c929a39dd6e2b807d29f198f581779e8c92567e07">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr></tbody></table>

</details>

___

### CODE REVIEW

File: runtimes/universal/models/gguf_language_model.py [hunk 1]
```diff
@@ -802,6 +802,28 @@ def _prepare_messages_with_tools(
...
+              messages and should not be injected again during generation.
+        """
+        if not tools:
+            return messages, False
+
+        # If the model supports native tool rendering, keep message shape unchanged.
+        if self._render_with_jinja2(messages, tools) is not None:
+            return messages, False
+
+        return self._prepare_messages_with_tools(messages, tools, tool_choice), True
```
[the remark this answers is not in the document]
Comment (rachmlenig) [reply runtimes/universal/models/gguf_language_model.py]: Good catch — this is valid and now covered.

  Added focused tests for all three branches in `runtimes/universal/tests/test_gguf_model.py`:
  - no tools
  - native Jinja2 render path
  - non-native tool injection path

  This keeps coverage tight to the changed logic without broad test churn.

File: runtimes/universal/utils/context_manager.py [hunk 2]
```diff
@@ -66,13 +66,25 @@ def from_context_size(
...
+        max_safe_margin = max(0, safe_n_ctx - 2)
+        safety_margin = min(safety_margin, max_safe_margin)
+
+        completion_target = max(1, max_completion_tokens)
+        available_after_safety = max(2, safe_n_ctx - safety_margin)
+
+        # Avoid hard-reserving large completion windows on tiny contexts.
+        adaptive_completion_cap = max(1, available_after_safety // 2)
+        reserved_completion = min(completion_target, adaptive_completion_cap)
+        max_prompt = max(1, safe_n_ctx - safety_margin - reserved_completion)
```
[the remark this answers is not in the document]
Comment (rachmlenig) [reply runtimes/universal/utils/context_manager.py:82]: Valid point. We kept the adaptive reservation for prompt budgeting safety on small contexts, but removed it as a hard cap on actual generation.

  `effective_max_tokens` is now computed from real remaining context after prompt accounting + safety margin in `runtimes/universal/routers/chat_completions/service.py`, so generation is no longer silently constrained by `reserved_completion`.

**Comment by rachmlenig** [comment]
Refreshing PR mergeability metadata; reopening immediately.
